src/resolve/resolved-dns-packet.c

```diff
@@ -2570,6 +2570,70 @@ bool dns_packet_equal(const DnsPacket *a, const DnsPacket *b) {
         return dns_packet_compare_func(a, b) == 0;
 }
 
+int dns_packet_ede_rcode(DnsPacket *p, uint16_t *ret_ede_rcode) {
+        assert(p);
+
+        bool has_ede = false;
+        uint16_t ede_rcode = _DNS_EDNS_OPT_INVALID;
+        const uint8_t *d;
+        size_t l;
+
+        if (!p->opt)
+                return false;
+
+        d = p->opt->opt.data;
+        l = p->opt->opt.data_size;
+
+        while (l > 0) {
+                uint16_t code, length;
+
+                if (l < 4U)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                               "EDNS0 variable part has invalid size.");
+
+                code = unaligned_read_be16(d);
+                length = unaligned_read_be16(d + 2);
+
+                if (l < 4U + length)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                               "Truncated option in EDNS0 variable part.");
+
+                if (code == DNS_EDNS_OPT_EXT_ERROR) {
+                        if (length < 2U)
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                "EDNS0 truncated EDE info code.");
+                        has_ede = true;
+                        ede_rcode = unaligned_read_be16(d+4);
+                        if (ede_rcode >= _DNS_EDNS_OPT_MAX_DEFINED)
+                                return -EINVAL;
+                        break;
+                }
+
+                d += 4U + length;
+                l -= 4U + length;
+        }
+
+        if (ret_ede_rcode)
+                *ret_ede_rcode = ede_rcode;
+
+        return has_ede;
```

Comment:
to simplify things I'd just return the parsed edercode as return value here. i.e. >= 0 means we found a valid ede rcode, and < 0 an error.